tools/publisher/src/package.rs

```diff
@@ -0,0 +1,404 @@
+    MissingVersion(PathBuf, String),
+    #[error("crate {0} has multiple versions: {1} and {2}")]
+    MultipleVersions(String, Version, Version),
+}
+
+/// Discovers all Cargo.toml files under the given path with a depth limit of 1.
+pub async fn discover_package_manifests(path: impl AsRef<Path>) -> Result<Vec<PathBuf>> {
+    let mut manifests = Vec::new();
+    let mut read_dir = fs::read_dir(path).await?;
+    while let Some(entry) = read_dir.next_entry().await? {
+        let package_path = entry.path();
+        if package_path.is_dir() {
+            let manifest_path = package_path.join("Cargo.toml");
+            if manifest_path.exists() {
+                manifests.push(manifest_path);
+            }
+        }
+    }
+    Ok(manifests)
+}
+
+/// Parses a semver version number and adds additional error context when parsing fails.
+pub fn parse_version(manifest_path: &Path, version: &str) -> Result<Version, Error> {
+    Version::parse(&version)
+        .map_err(|err| Error::InvalidCrateVersion(manifest_path.into(), version.into(), err.into()))
+}
+
+fn read_dependencies(path: &Path, dependencies: &DepsSet) -> Result<Vec<PackageHandle>> {
+    let mut result = Vec::new();
+    for (name, metadata) in dependencies {
+        match metadata {
+            Dependency::Simple(_) => {}
+            Dependency::Detailed(detailed) => {
+                if detailed.path.is_some() {
+                    let version = detailed
+                        .version
+                        .as_ref()
+                        .map(|version| parse_version(path, &version))
+                        .ok_or_else(|| Error::MissingVersion(path.into(), name.into()))??;
+                    result.push(PackageHandle::new(name, version));
+                }
+            }
+        }
+    }
+    Ok(result)
+}
+
+fn read_package(path: &Path, manifest_bytes: &[u8]) -> Result<Package> {
+    let manifest = Manifest::from_slice(manifest_bytes)
+        .with_context(|| format!("failed to load package manifest for {:?}", path))?;
+    let package = manifest
+        .package
+        .ok_or_else(|| Error::InvalidManifest(path.into()))
+        .context("crate manifest doesn't have a `[package]` section")?;
+    let name = package.name;
+    let version = parse_version(path, &package.version)?;
+    let handle = PackageHandle { name, version };
+
+    let mut local_dependencies = BTreeSet::new();
+    local_dependencies.extend(read_dependencies(path, &manifest.dependencies)?.into_iter());
+    local_dependencies.extend(read_dependencies(path, &manifest.dev_dependencies)?.into_iter());
+    local_dependencies.extend(read_dependencies(path, &manifest.build_dependencies)?.into_iter());
+    Ok(Package::new(handle, path, local_dependencies))
+}
+
+/// Validates that all of the publishable crates use consistent version numbers
+/// across all of their local dependencies.
+fn validate_packages(packages: &Vec<Package>) -> Result<()> {
```

Comment:
clippy won't let you do this, `&[Package]`

Reply:
Oh, I should run clippy on this... and also add it to CI :smile: